pyvrp/cpp/SubPopulation.cpp

```diff
@@ -0,0 +1,156 @@
+
+    for (auto &other : items)  // update distance to other individuals
+    {
+        auto const div = divOp(data, *individual, *other.individual);
+        auto cmp = [](auto &elem, auto &value) { return elem.first < value; };
+
+        auto &oProx = other.proximity;
+        auto place = std::lower_bound(oProx.begin(), oProx.end(), div, cmp);
+        oProx.emplace(place, div, individual);
+
+        auto &iProx = item.proximity;
+        place = std::lower_bound(iProx.begin(), iProx.end(), div, cmp);
+        iProx.emplace(place, div, other.individual);
+    }
+
+    items.push_back(item);
+    updateFitness();
+
+    if (size() > params.maxPopSize())
+        purge();
+}
+
+size_t SubPopulation::size() const { return items.size(); }
+
+SubPopulation::Item const &SubPopulation::operator[](size_t idx) const
+{
+    return items[idx];
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cbegin() const
+{
+    return items.cbegin();
+}
+
+std::vector<SubPopulation::Item>::const_iterator SubPopulation::cend() const
+{
+    return items.cend();
+}
+
+void SubPopulation::remove(
+    std::vector<SubPopulation::Item>::iterator const &iterator)
+{
+    for (auto &[params, individual, fitness, proximity] : items)
+        // Remove individual from other proximities.
+        for (size_t idx = 0; idx != proximity.size(); ++idx)
+            if (proximity[idx].second == iterator->individual)
+            {
+                proximity.erase(proximity.begin() + idx);
+                break;
+            }
+
+    delete iterator->individual;  // dispose of manually allocated memory
+    items.erase(iterator);        // before the item is removed.
+}
+
+void SubPopulation::purge()
+{
+    while (size() > params.minPopSize)
+    {
+        // Remove duplicates from the subpopulation (if they exist)
+        auto const pred = [&](auto &iterator)
+        {
+            return !iterator.proximity.empty()
+                   && *iterator.proximity[0].second == *iterator.individual;
+        };
+
+        auto const duplicate = std::find_if(items.begin(), items.end(), pred);
+
+        if (duplicate == items.end())  // there are no more duplicates
+            break;
+
+        remove(duplicate);
+    }
+
+    while (size() > params.minPopSize)
+    {
+        // Remove the worst individual (worst in terms of biased fitness)
+        updateFitness();
+
+        auto const worstFitness = std::max_element(
+            items.begin(),
+            items.end(),
+            [](auto const &a, auto const &b) { return a.fitness < b.fitness; });
+
+        remove(worstFitness);
+    }
+}
+
+void SubPopulation::updateFitness()
+{
+    if (items.empty())
+        return;
+
+    std::vector<size_t> byCost(size());
+    std::iota(byCost.begin(), byCost.end(), 0);
+
+    std::stable_sort(
+        byCost.begin(),
+        byCost.end(),
+        [&](size_t a, size_t b)
+        { return items[a].individual->cost() < items[b].individual->cost(); });
```

Comment:
Note that this will compute cost per individual multiple times. I think it's probably fast enough currently but good to be aware of if that were to change.